File: app/src/main/java/com/bhashasetu/app/database/AppDatabase.java
package com.bhashasetu.app.database;

import android.content.Context;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.migration.Migration;
import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.room.TypeConverters;

//Achievement System Imports (Updated for Consolidation)
// import com.bhashasetu.app.model.Achievement; //Legacy Achievement - removed to avoid import conflict
// import com.bhashasetu.app.data.model.Achievement; //Modern Achievement - removed to avoid import conflict

// Core Entity Imports
import com.bhashasetu.app.model.LessonWord;
import com.bhashasetu.app.model.User;
import com.bhashasetu.app.model.Word;
import com.bhashasetu.app.model.WordCategory;
import com.bhashasetu.app.model.Lesson;
import com.bhashasetu.app.model.UserProgress;
import com.bhashasetu.app.model.GameScore;
import com.bhashasetu.app.pronunciation.PronunciationProgressStats;
import com.bhashasetu.app.model.PronunciationSession;

// Gamification Imports (NO Achievement conflicts here!)
import com.bhashasetu.app.model.gamification.Badge;
import com.bhashasetu.app.model.gamification.UserLevel;
import com.bhashasetu.app.model.gamification.UserPoints;
import com.bhashasetu.app.model.gamification.UserStats;

// Pronunciation Imports
import com.bhashasetu.app.pronunciation.PronunciationScore;
import com.bhashasetu.app.pronunciation.PronunciationScoreDao;
import com.bhashasetu.app.pronunciation.PronunciationSessionDao;

// Required DAO Imports
import com.bhashasetu.app.database.AchievementDao;
import com.bhashasetu.app.database.ModernAchievementDao;
import com.bhashasetu.app.database.WordDao;
import com.bhashasetu.app.database.WordCategoryDao;
import com.bhashasetu.app.database.UserDao;
import com.bhashasetu.app.database.UserProgressDao;
import com.bhashasetu.app.database.GameScoreDao;
import com.bhashasetu.app.database.BadgeDao;
import com.bhashasetu.app.database.UserLevelDao;
import com.bhashasetu.app.database.UserPointsDao;
import com.bhashasetu.app.database.UserStatsDao;
import com.bhashasetu.app.database.Converters;

// Modern DAO Imports
import com.bhashasetu.app.data.dao.UserGoalDao;
import com.bhashasetu.app.data.dao.LessonDao;

// Migration Import
import com.bhashasetu.app.database.AchievementMigrations;

/**
 * Main Room database class for the application.
 * ✅ ROOM ERROR FIX: Added missing entities and fixed syntax errors
 * Consolidated Achievement System - Two Clean Tables -
 * 1. legacy_achievements (backward compatibility)
 * 2. achievements (modern bilingual system)
 * Defines all entities and provides DAOs for database access.
 */
@Database(entities = {
        // Core Entities
        User.class,
        Word.class,
//        com.bhashasetu.app.model.Word.class,  // ✅ Use the fixed one
        WordCategory.class,
        Lesson.class,
        UserProgress.class,
        GameScore.class,
        LessonWord.class,
        PronunciationProgressStats.class,
        // DON'T include com.bhashasetu.app.data.model.Word

        // Achievement Systems (Consolidated - No More Conflicts)
        com.bhashasetu.app.model.Achievement.class, //Legacy achievement - legacy_achievements
        com.bhashasetu.app.data.model.Achievement.class, //Modern achievement - achievements table

        // Gamification (Achievement conflicts removed)
        Badge.class,
        UserLevel.class,
        UserPoints.class,
        UserStats.class,

        // Pronunciation
        PronunciationSession.class,
        PronunciationScore.class,

        // ✅ ROOM ERROR FIX: Add missing modern entities for lesson_words table and others
        com.bhashasetu.app.data.model.UserGoal.class,
        com.bhashasetu.app.data.model.Lesson.class,
        com.bhashasetu.app.data.model.LessonWord.class,
        com.bhashasetu.app.data.model.Category.class,
        com.bhashasetu.app.data.model.Quiz.class,
        com.bhashasetu.app.data.model.QuizQuestion.class,
        com.bhashasetu.app.data.model.StudySession.class,
        com.bhashasetu.app.data.model.DailyStreak.class,
        com.bhashasetu.app.data.model.AppSettings.class,
        com.bhashasetu.app.data.model.UserProgress.class,
        com.bhashasetu.app.data.model.Word.class,
        com.bhashasetu.app.data.model.WordCategoryCrossRef.class
},
        version = 8, exportSchema = true)  // ✅ INCREMENT VERSION FOR SCHEMA CHANGES
@TypeConverters(Converters.class)

public abstract class AppDatabase extends RoomDatabase {

    private static final String DATABASE_NAME = "english_hindi_db";
    private static volatile AppDatabase INSTANCE;

    // DAOs
    // ACHIEVEMENT SYSTEM DAOs (Consolidated)
    public abstract AchievementDao achievementDao(); //Legacy Achievement Dao
    public abstract ModernAchievementDao modernAchievementDao(); // Modern Achievement Dao

    // CORE SYSTEM DAOs
    public abstract WordDao wordDao();
    public abstract WordCategoryDao wordCategoryDao();
    public abstract UserDao userDao();
    public abstract UserProgressDao userProgressDao();
    public abstract GameScoreDao gameScoreDao();

    // GAMIFICATION DAOs (Achievement conflicts removed)
    public abstract BadgeDao badgeDao();
    public abstract UserLevelDao userLevelDao();
    public abstract UserPointsDao userPointsDao();
    public abstract UserStatsDao userStatsDao();

    // PRONUNCIATION DAOs
    public abstract PronunciationSessionDao pronunciationSessionDao();
    public abstract PronunciationScoreDao pronunciationScoreDao();
    public abstract PronunciationProgressStatsDao pronunciationProgressStatsDao();

    // ✅ MODERN SYSTEM DAOs
    public abstract UserGoalDao userGoalDao();
    public abstract LessonDao lessonDao();
    public abstract appSettingsDao appSettingsDao();

    /**
     * Get singleton instance of the database with achievement consolidation migrations.
     * @param context Application context
     * @return The AppDatabase instance
     */
    public static AppDatabase getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (AppDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(
                                    context.getApplicationContext(),
                                    AppDatabase.class,
                                    DATABASE_NAME)
                            // ACHIEVEMENT CONSOLIDATION MIGRATIONS
                            .addMigrations(AchievementMigrations.MIGRATION_3_4) // We'll create this later
                            .fallbackToDestructiveMigration() // Keep as fallback -Use with caution in production
                            .build();
                }
            }
        }
        return INSTANCE;
    }

    // Migration from version 3 to 4
    private static final Migration MIGRATION_3_4 = new Migration(3, 4) {
        @Override
        public void migrate(SupportSQLiteDatabase database) {
            // Handle table name changes to resolve conflicts
            // 1. Rename old words table if it exists
            database.execSQL("ALTER TABLE words RENAME TO words_old");

            // 2. Create new vocabulary_words table
            database.execSQL("CREATE TABLE IF NOT EXISTS `vocabulary_words` (" +
                    "`id` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, " +
                    "`englishText` TEXT, " +
                    "`localText` TEXT, " +
                    "`pronunciation` TEXT, " +
                    "`phonetic` TEXT, " +
                    "`audioUrl` TEXT, " +
                    "`imageUrl` TEXT, " +
                    "`definition` TEXT, " +
                    "`exampleSentence` TEXT, " +
                    "`exampleTranslation` TEXT, " +
                    "`categoryId` INTEGER NOT NULL, " +
                    "`lessonId` INTEGER NOT NULL, " +
                    "`difficultyLevel` INTEGER NOT NULL, " +
                    "`orderInLesson` INTEGER NOT NULL, " +
                    "`isActive` INTEGER NOT NULL, " +
                    "`createdAt` INTEGER NOT NULL, " +
                    "`updatedAt` INTEGER NOT NULL, " +
                    "`partOfSpeech` TEXT, " +
                    "`synonyms` TEXT, " +
                    "`antonyms` TEXT, " +
                    "`frequency` INTEGER NOT NULL, " +
                    "`languageCode` TEXT, " +
                    "FOREIGN KEY(`categoryId`) REFERENCES `word_categories`(`id`) ON UPDATE NO ACTION ON DELETE CASCADE)");

            // 3. Create indexes for vocabulary_words
            database.execSQL("CREATE INDEX IF NOT EXISTS `index_vocabulary_words_categoryId` ON `vocabulary_words` (`categoryId`)");
            database.execSQL("CREATE INDEX IF NOT EXISTS `index_vocabulary_words_lessonId` ON `vocabulary_words` (`lessonId`)");
            database.execSQL("CREATE INDEX IF NOT EXISTS `index_vocabulary_words_englishText` ON `vocabulary_words` (`englishText`)");

            // 4. Migrate data from old table to new table
            database.execSQL("INSERT INTO vocabulary_words (id, englishText, localText, pronunciation, definition, categoryId, lessonId, difficultyLevel, orderInLesson, isActive, createdAt, updatedAt, frequency) " +
                    "SELECT id, " +
                    "COALESCE(englishText, text, word) as englishText, " +
                    "COALESCE(localText, translation, localTranslation) as localText, " +
                    "pronunciation, " +
                    "COALESCE(definition, meaning) as definition, " +
                    "COALESCE(categoryId, category_id, 1) as categoryId, " +
                    "COALESCE(lessonId, lesson_id, 1) as lessonId, " +
                    "COALESCE(difficultyLevel, difficulty, 1) as difficultyLevel, " +
                    "COALESCE(orderInLesson, order_in_lesson, 0) as orderInLesson, " +
                    "COALESCE(isActive, is_active, 1) as isActive, " +
                    "COALESCE(createdAt, created_at, strftime('%s', 'now') * 1000) as createdAt, " +
                    "COALESCE(updatedAt, updated_at, strftime('%s', 'now') * 1000) as updatedAt, " +
                    "COALESCE(frequency, 1) as frequency " +
                    "FROM words_old");

            // 5. Drop old table
            database.execSQL("DROP TABLE words_old");

            // 6. Handle user_progress table rename
            database.execSQL("ALTER TABLE user_progress RENAME TO user_progress_old");

            // 7. Create new user_learning_progress table
            database.execSQL("CREATE TABLE IF NOT EXISTS `user_learning_progress` (" +
                    "`id` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, " +
                    "`userId` INTEGER NOT NULL, " +
                    "`wordId` INTEGER NOT NULL, " +
                    "`lessonId` INTEGER NOT NULL, " +
                    "`masteryLevel` INTEGER NOT NULL, " +
                    "`correctAttempts` INTEGER NOT NULL, " +
                    "`totalAttempts` INTEGER NOT NULL, " +
                    "`accuracy` REAL NOT NULL, " +
                    "`firstLearnedAt` INTEGER NOT NULL, " +
                    "`lastReviewedAt` INTEGER NOT NULL, " +
                    "`nextReviewAt` INTEGER NOT NULL, " +
                    "`isCompleted` INTEGER NOT NULL, " +
                    "`isMastered` INTEGER NOT NULL, " +
                    "`streakCount` INTEGER NOT NULL, " +
                    "`reviewCount` INTEGER NOT NULL, " +
                    "`learningStage` TEXT, " +
                    "`totalTimeSpent` INTEGER NOT NULL, " +
                    "`difficulty` TEXT, " +
                    "FOREIGN KEY(`userId`) REFERENCES `users`(`id`) ON UPDATE NO ACTION ON DELETE CASCADE, " +
                    "FOREIGN KEY(`wordId`) REFERENCES `vocabulary_words`(`id`) ON UPDATE NO ACTION ON DELETE CASCADE)");

            // 8. Create indexes for user_learning_progress
            database.execSQL("CREATE INDEX IF NOT EXISTS `index_user_learning_progress_userId` ON `user_learning_progress` (`userId`)");
            database.execSQL("CREATE INDEX IF NOT EXISTS `index_user_learning_progress_wordId` ON `user_learning_progress` (`wordId`)");
            database.execSQL("CREATE INDEX IF NOT EXISTS `index_user_learning_progress_lessonId` ON `user_learning_progress` (`lessonId`)");

            // 9. Migrate user progress data
            database.execSQL("INSERT INTO user_learning_progress (id, userId, wordId, lessonId, masteryLevel, correctAttempts, totalAttempts, accuracy, firstLearnedAt, lastReviewedAt, nextReviewAt, isCompleted, isMastered, streakCount, reviewCount, learningStage, totalTimeSpent, difficulty) " +
                    "SELECT id, " +
                    "userId, " +
                    "wordId, " +
                    "COALESCE(lessonId, 1) as lessonId, " +
                    "COALESCE(masteryLevel, progress, 0) as masteryLevel, " +
                    "COALESCE(correctAttempts, correct_attempts, 0) as correctAttempts, " +
                    "COALESCE(totalAttempts, total_attempts, 0) as totalAttempts, " +
                    "COALESCE(accuracy, 0.0) as accuracy, " +
                    "COALESCE(firstLearnedAt, first_learned_at, created_at, strftime('%s', 'now') * 1000) as firstLearnedAt, " +
                    "COALESCE(lastReviewedAt, last_reviewed_at, updated_at, strftime('%s', 'now') * 1000) as lastReviewedAt, " +
                    "COALESCE(nextReviewAt, strftime('%s', 'now') * 1000 + 86400000) as nextReviewAt, " +
                    "COALESCE(isCompleted, is_completed, 0) as isCompleted, " +
                    "COALESCE(isMastered, is_mastered, 0) as isMastered, " +
                    "COALESCE(streakCount, streak_count, 0) as streakCount, " +
                    "COALESCE(reviewCount, review_count, 0) as reviewCount, " +
                    "COALESCE(learningStage, 'learning') as learningStage, " +
                    "COALESCE(totalTimeSpent, total_time_spent, 0) as totalTimeSpent, " +
                    "COALESCE(difficulty, 'beginner') as difficulty " +
                    "FROM user_progress_old");

            // 10. Drop old user_progress table
            database.execSQL("DROP TABLE user_progress_old");

            // 11. Add missing indexes for game_scores
            database.execSQL("CREATE INDEX IF NOT EXISTS `index_game_scores_userId` ON `game_scores` (`userId`)");
            database.execSQL("CREATE INDEX IF NOT EXISTS `index_game_scores_gameType` ON `game_scores` (`gameType`)");
        }
    };
}

File: app/src/main/java/com/bhashasetu/app/data/db/AppDatabase.kt
package com.bhashasetu.app.data.db

import android.content.Context
import androidx.room.*
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

// Data Access Objects
import com.bhashasetu.app.data.dao.*
import com.bhashasetu.app.database.*

// Modern Kotlin Models
import com.bhashasetu.app.data.model.*

// Legacy Java Models (for backward compatibility)
import com.bhashasetu.app.model.User
import com.bhashasetu.app.model.Word as LegacyWord
import com.bhashasetu.app.model.WordCategory
import com.bhashasetu.app.model.Lesson as LegacyLesson
import com.bhashasetu.app.model.UserProgress as LegacyUserProgress
import com.bhashasetu.app.model.GameScore
import com.bhashasetu.app.model.LessonWord as LegacyLessonWord
import com.bhashasetu.app.model.PronunciationSession
import com.bhashasetu.app.model.Achievement as LegacyAchievement

// Gamification Models
import com.bhashasetu.app.model.gamification.Badge
import com.bhashasetu.app.model.gamification.UserLevel
import com.bhashasetu.app.model.gamification.UserPoints
import com.bhashasetu.app.model.gamification.UserStats

// Pronunciation Models
import com.bhashasetu.app.pronunciation.PronunciationProgressStats
import com.bhashasetu.app.pronunciation.PronunciationScore

// Converters
import com.bhashasetu.app.data.util.DateConverter
import com.bhashasetu.app.database.Converters

/**
 * Comprehensive AppDatabase that consolidates all functionality
 *
 * This database includes:
 * - Modern Kotlin entities with coroutines support
 * - Legacy Java entities for backward compatibility
 * - Proper migration system
 * - Achievement system (both legacy and modern)
 * - Gamification features
 * - Pronunciation tracking
 * - Offline content management
 * - User progress and analytics
 */
@Database(
    entities = [
        // Modern Kotlin Entities (Primary)
        Word::class,
        Category::class,
        WordCategoryCrossRef::class,
        UserProgress::class,
        Lesson::class,
        LessonWord::class,
        Quiz::class,
        QuizQuestion::class,
        Achievement::class,
        AppSettings::class,
        StudySession::class,
        UserGoal::class,
        DailyStreak::class,

        // Legacy Java Entities (Backward Compatibility)
        User::class,
        LegacyWord::class,
        WordCategory::class,
        LegacyLesson::class,
        LegacyUserProgress::class,
        GameScore::class,
        LegacyLessonWord::class,
        PronunciationProgressStats::class,
        LegacyAchievement::class,

        // Gamification Entities
        Badge::class,
        UserLevel::class,
        UserPoints::class,
        UserStats::class,

        // Pronunciation Entities
        PronunciationSession::class,
        PronunciationScore::class
    ],
    version = 10,
    exportSchema = true,
    autoMigrations = [
        AutoMigration(from = 8, to = 9),
        AutoMigration(from = 9, to = 10)
    ]
)
@TypeConverters(DateConverter::class, Converters::class)
abstract class AppDatabase : RoomDatabase() {

    // Modern Kotlin DAOs (Primary)
    abstract fun wordDao(): WordDao
    abstract fun categoryDao(): CategoryDao
    abstract fun userProgressDao(): UserProgressDao
    abstract fun lessonDao(): LessonDao
    abstract fun quizDao(): QuizDao
    abstract fun achievementDao(): AchievementDao
    abstract fun settingsDao(): SettingsDao
    abstract fun studySessionDao(): StudySessionDao
    abstract fun userGoalDao(): UserGoalDao
    abstract fun dailyStreakDao(): DailyStreakDao
    abstract fun appSettingsDao(): AppSettingsDao

    // Legacy Java DAOs (Backward Compatibility)
    abstract fun legacyWordDao(): com.bhashasetu.app.database.WordDao
    abstract fun wordCategoryDao(): WordCategoryDao
    abstract fun userDao(): UserDao
    abstract fun legacyUserProgressDao(): com.bhashasetu.app.database.UserProgressDao
    abstract fun gameScoreDao(): GameScoreDao
    abstract fun legacyAchievementDao(): com.bhashasetu.app.database.AchievementDao
    abstract fun modernAchievementDao(): ModernAchievementDao

    // Gamification DAOs
    abstract fun badgeDao(): BadgeDao
    abstract fun userLevelDao(): UserLevelDao
    abstract fun userPointsDao(): UserPointsDao
    abstract fun userStatsDao(): UserStatsDao

    // Pronunciation DAOs
    abstract fun pronunciationSessionDao(): com.bhashasetu.app.pronunciation.PronunciationSessionDao
    abstract fun pronunciationScoreDao(): com.bhashasetu.app.pronunciation.PronunciationScoreDao
    abstract fun pronunciationProgressStatsDao(): PronunciationProgressStatsDao

    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null

        private const val DATABASE_NAME = "bhashasetu_comprehensive_db"

        /**
         * Get database instance with proper initialization
         */
        fun getDatabase(
            context: Context,
            scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
        ): AppDatabase {
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    DATABASE_NAME
                )
                    .addMigrations(
                        MIGRATION_7_8,
                        MIGRATION_8_9,
                        MIGRATION_9_10
                    )
                    .addCallback(AppDatabaseCallback(scope))
                    .fallbackToDestructiveMigration() // Only for development
                    .build()
                INSTANCE = instance
                instance
            }
        }

        /**
         * Alternative method for Java compatibility
         */
        fun getInstance(context: Context): AppDatabase {
            return getDatabase(context)
        }

        /**
         * Migration from version 7 to 8 - Achievement system consolidation
         */
        private val MIGRATION_7_8 = object : Migration(7, 8) {
            override fun migrate(database: SupportSQLiteDatabase) {
                // Create new achievement tables to resolve conflicts
                database.execSQL("""
                    CREATE TABLE IF NOT EXISTS `modern_achievements` (
                        `id` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
                        `titleEn` TEXT NOT NULL,
                        `titleHi` TEXT NOT NULL,
                        `descriptionEn` TEXT NOT NULL,
                        `descriptionHi` TEXT NOT NULL,
                        `iconName` TEXT NOT NULL,
                        `type` TEXT NOT NULL,
                        `targetValue` INTEGER NOT NULL,
                        `currentValue` INTEGER NOT NULL DEFAULT 0,
                        `isUnlocked` INTEGER NOT NULL DEFAULT 0,
                        `unlockedAt` INTEGER,
                        `category` TEXT NOT NULL,
                        `difficulty` TEXT NOT NULL,
                        `xpReward` INTEGER NOT NULL DEFAULT 0,
                        `createdAt` INTEGER NOT NULL,
                        `updatedAt` INTEGER NOT NULL
                    )
                """.trimIndent())

                // Migrate existing achievements if they exist
                database.execSQL("""
                    INSERT INTO modern_achievements (titleEn, titleHi, descriptionEn, descriptionHi, iconName, type, targetValue, category, difficulty, xpReward, createdAt, updatedAt)
                    SELECT title, title, description, description, 'default', 'progress', 100, 'general', 'easy', 50, 
                           strftime('%s', 'now') * 1000, strftime('%s', 'now') * 1000
                    FROM achievements WHERE achievements.id IS NOT NULL
                """.trimIndent())
            }
        }

        /**
         * Migration from version 8 to 9 - Enhanced word tracking
         */
        private val MIGRATION_8_9 = object : Migration(8, 9) {
            override fun migrate(database: SupportSQLiteDatabase) {
                // Add new columns for better word tracking
                database.execSQL("ALTER TABLE words ADD COLUMN `difficultyScore` REAL NOT NULL DEFAULT 0.0")
                database.execSQL("ALTER TABLE words ADD COLUMN `lastPracticed` INTEGER DEFAULT NULL")
                database.execSQL("ALTER TABLE words ADD COLUMN `masteryLevel` INTEGER NOT NULL DEFAULT 0")
                database.execSQL("ALTER TABLE words ADD COLUMN `practiceCount` INTEGER NOT NULL DEFAULT 0")
                database.execSQL("ALTER TABLE words ADD COLUMN `correctAnswers` INTEGER NOT NULL DEFAULT 0")
                database.execSQL("ALTER TABLE words ADD COLUMN `totalAnswers` INTEGER NOT NULL DEFAULT 0")
            }
        }

        /**
         * Migration from version 9 to 10 - Settings and streaks
         */
        private val MIGRATION_9_10 = object : Migration(9, 10) {
            override fun migrate(database: SupportSQLiteDatabase) {
                // Create daily streaks table
                database.execSQL("""
                    CREATE TABLE IF NOT EXISTS `daily_streaks` (
                        `id` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
                        `date` TEXT NOT NULL,
                        `studyTimeMinutes` INTEGER NOT NULL DEFAULT 0,
                        `wordsLearned` INTEGER NOT NULL DEFAULT 0,
                        `lessonsCompleted` INTEGER NOT NULL DEFAULT 0,
                        `quizzesTaken` INTEGER NOT NULL DEFAULT 0,
                        `pronunciationPracticed` INTEGER NOT NULL DEFAULT 0,
                        `streakCount` INTEGER NOT NULL DEFAULT 0,
                        `isGoalMet` INTEGER NOT NULL DEFAULT 0,
                        `createdAt` INTEGER NOT NULL
                    )
                """.trimIndent())

                // Create unique index on date
                database.execSQL("CREATE UNIQUE INDEX IF NOT EXISTS `index_daily_streaks_date` ON `daily_streaks` (`date`)")

                // Enhance app settings
                database.execSQL("ALTER TABLE app_settings ADD COLUMN `achievementNotifications` INTEGER NOT NULL DEFAULT 1")
                database.execSQL("ALTER TABLE app_settings ADD COLUMN `difficultyLevel` TEXT NOT NULL DEFAULT 'Beginner'")
                database.execSQL("ALTER TABLE app_settings ADD COLUMN `dailyGoalMinutes` INTEGER NOT NULL DEFAULT 15")
                database.execSQL("ALTER TABLE app_settings ADD COLUMN `offlineModeEnabled` INTEGER NOT NULL DEFAULT 0")
                database.execSQL("ALTER TABLE app_settings ADD COLUMN `analyticsEnabled` INTEGER NOT NULL DEFAULT 1")
            }
        }

        /**
         * Database callback for initialization
         */
        private class AppDatabaseCallback(
            private val scope: CoroutineScope
        ) : RoomDatabase.Callback() {

            override fun onCreate(db: SupportSQLiteDatabase) {
                super.onCreate(db)
                INSTANCE?.let { database ->
                    scope.launch {
                        prePopulateDatabase(database)
                    }
                }
            }

            override fun onOpen(db: SupportSQLiteDatabase) {
                super.onOpen(db)
                // Ensure foreign key constraints are enabled
                db.execSQL("PRAGMA foreign_keys=ON;")
            }
        }

        /**
         * Pre-populate database with initial data
         */
        private suspend fun prePopulateDatabase(database: AppDatabase) {
            try {
                // Add default categories
                val categoryDao = database.categoryDao()
                val categories = listOf(
                    Category(
                        id = 1,
                        name = "Basic Words",
                        nameHindi = "बुनियादी शब्द",
                        description = "Essential vocabulary for beginners",
                        descriptionHindi = "शुरुआती लोगों के लिए आवश्यक शब्दावली",
                        colorCode = "#FF6B6B",
                        iconName = "basic"
                    ),
                    Category(
                        id = 2,
                        name = "Food & Drink",
                        nameHindi = "खाना और पेय",
                        description = "Food and beverage vocabulary",
                        descriptionHindi = "खाने-पीने की शब्दावली",
                        colorCode = "#4ECDC4",
                        iconName = "food"
                    ),
                    Category(
                        id = 3,
                        name = "Numbers",
                        nameHindi = "संख्याएं",
                        description = "Numbers and counting",
                        descriptionHindi = "संख्या और गिनती",
                        colorCode = "#45B7D1",
                        iconName = "numbers"
                    ),
                    Category(
                        id = 4,
                        name = "Family",
                        nameHindi = "परिवार",
                        description = "Family relationships and terms",
                        descriptionHindi = "पारिवारिक रिश्ते और शब्द",
                        colorCode = "#F7DC6F",
                        iconName = "family"
                    ),
                    Category(
                        id = 5,
                        name = "Colors",
                        nameHindi = "रंग",
                        description = "Colors and shades",
                        descriptionHindi = "रंग और उनके शेड्स",
                        colorCode = "#BB8FCE",
                        iconName = "colors"
                    )
                )
                categoryDao.insertAll(categories)

                // Add default app settings
                val settingsDao = database.appSettingsDao()
                val defaultSettings = AppSettings(
                    id = 1,
                    language = "en",
                    notificationsEnabled = true,
                    dailyReminderTime = "09:00",
                    darkModeEnabled = false,
                    soundEnabled = true,
                    vibrationEnabled = true,
                    difficultyLevel = "Beginner",
                    dailyGoalMinutes = 15,
                    offlineModeEnabled = false,
                    analyticsEnabled = true,
                    achievementNotifications = true
                )
                settingsDao.insert(defaultSettings)

                // Add sample words for each category
                val wordDao = database.wordDao()
                val sampleWords = listOf(
                    // Basic Words
                    Word(
                        id = 1,
                        english = "Hello",
                        hindi = "नमस्ते",
                        pronunciation = "namaste",
                        phonetic = "/nə-məs-ˈtā/",
                        definition = "A greeting used when meeting someone",
                        definitionHindi = "किसी से मिलते समय प्रयोग होने वाला अभिवादन",
                        exampleSentence = "Hello, how are you?",
                        exampleSentenceHindi = "नमस्ते, आप कैसे हैं?",
                        categoryId = 1,
                        difficultyLevel = 1,
                        isActive = true
                    ),
                    Word(
                        id = 2,
                        english = "Thank you",
                        hindi = "धन्यवाद",
                        pronunciation = "dhanyawad",
                        phonetic = "/dʰən-jə-ˈwɑːd/",
                        definition = "An expression of gratitude",
                        definitionHindi = "कृतज्ञता का प्रकाशन",
                        exampleSentence = "Thank you for your help",
                        exampleSentenceHindi = "आपकी सहायता के लिए धन्यवाद",
                        categoryId = 1,
                        difficultyLevel = 1,
                        isActive = true
                    ),
                    // Food Words
                    Word(
                        id = 3,
                        english = "Water",
                        hindi = "पानी",
                        pronunciation = "paani",
                        phonetic = "/pɑː-niː/",
                        definition = "A clear liquid essential for life",
                        definitionHindi = "जीवन के लिए आवश्यक एक स्वच्छ तरल",
                        exampleSentence = "I drink water every day",
                        exampleSentenceHindi = "मैं हर दिन पानी पीता हूं",
                        categoryId = 2,
                        difficultyLevel = 1,
                        isActive = true
                    )
                )
                wordDao.insertAll(sampleWords)

                // Add initial achievements
                val achievementDao = database.achievementDao()
                val initialAchievements = listOf(
                    Achievement(
                        id = 1,
                        titleEn = "First Steps",
                        titleHi = "पहले कदम",
                        descriptionEn = "Learn your first 5 words",
                        descriptionHi = "अपने पहले 5 शब्द सीखें",
                        iconName = "first_steps",
                        type = "word_count",
                        targetValue = 5,
                        category = "learning",
                        difficulty = "easy",
                        xpReward = 100
                    ),
                    Achievement(
                        id = 2,
                        titleEn = "Practice Makes Perfect",
                        titleHi = "अभ्यास से सफलता",
                        descriptionEn = "Complete 7 days of practice",
                        descriptionHi = "7 दिन अभ्यास पूरा करें",
                        iconName = "practice_streak",
                        type = "streak",
                        targetValue = 7,
                        category = "consistency",
                        difficulty = "medium",
                        xpReward = 250
                    )
                )
                achievementDao.insertAll(initialAchievements)

            } catch (e: Exception) {
                // Log error but don't crash the app
                android.util.Log.e("AppDatabase", "Error pre-populating database", e)
            }
        }

        /**
         * Clear all data (for testing or reset)
         */
        suspend fun clearAllData(database: AppDatabase) {
            database.clearAllTables()
        }

        /**
         * Export database (for backup)
         */
        fun exportDatabase(context: Context): String? {
            return try {
                val dbFile = context.getDatabasePath(DATABASE_NAME)
                dbFile.absolutePath
            } catch (e: Exception) {
                null
            }
        }
    }
}